pkg/auth/tls.go

```diff
@@ -54,3 +57,44 @@ func (p *TLSAuthProvider) GetTLSCertificate() (*tls.Certificate, error) {
 	cert, err := tls.LoadX509KeyPair(p.certificatePath, p.privateKeyPath)
 	return &cert, err
 }
+
+func (p *TLSAuthProvider) AddAuthParams(client *http.Client, req *http.Request) {
```

Comment:
The logic here is not adding auth params. You are initializing a tls transport. I don't think we should call it `AddAuthParams`.

Also `req` is not used in this method.

Reply:
Java client implementation already provides a good abstraction for different type of authentication implementation. Please check them out.

I don't think the abstraction here is a good abstraction.

1) initializing a transport should be done during initializing auth provider. we should not do it in a method that will be called for each request. You can try avoid initializing transport each time this method is called, by adding a check logic. It is a hack. but the abstraction is wrong.

2) for a method that will be called for each request, `http.client` is not needed. The java client already provides a good abstraction. 

```
/**
     * For mutual authentication, This method use passed in `data` to evaluate and challenge,
     * then returns null if authentication has completed;
     * returns authenticated data back to server side, if authentication has not completed.
     *
     * <p>Mainly used for mutual authentication like sasl.
     */
    default AuthData authenticate(AuthData data) throws AuthenticationException {
        byte[] bytes = (hasDataFromCommand() ? this.getCommandData() : "").getBytes(UTF_8);
        return AuthData.of(bytes);
    }
```

you can just add a method `authenticate(req *http.Request) (*http.Request)` in the auth provider. It is a clear abstraction.